### PR TITLE
Disallow untyped calls in mypy check

### PR DESCRIPTION
**Pull request by zundertj**
Fixes mypy "disallow_untyped_calls" in #4044 .

### CODE REVIEW

**Comment by zundertj** [comment]
Similar issue as in #4052 with numpy. The Python 3.7 CI run installs numpy 1.21.6, as numpy 1.22 and beyond drop Python 3.7 support (https://numpy.org/devdocs/release/1.22.0-notes.html). This means that there is no way to satisfy type annotations on older numpy, which is not fully type annotated and therefore triggers mypy's `no-untyped-call`, and at the same time also not use `typing.cast` or `type ignore` to satisfy the Python 3.10 run. So we have to make a choice here, do we want to please mypy on all possible Python versions we support, or are we happy to target the latest version(s) only? Not arguing for dropping Python 3.7 altogether, it seems it is from a testing perspective and the ability to use future imports, to not be material.

File: py-polars/tests/test_df.py [hunk 1]
```diff
@@ -1741,10 +1741,10 @@ def name_generator() -> Iterator[str]:
 
 def test_extension() -> None:
     class Foo:
-        def __init__(self, value):  # type: ignore
+        def __init__(self, value: Any) -> None:
```
Comment (stinodego) [inline py-polars/tests/test_df.py:1744]: Looking at the context, `value` can be an `int` type here. And in general, `__init__` methods should not specify their return type (it is always `-> None`), unless it has no other arguments (which would make it an untyped def). Explanation [here](https://github.com/python/mypy/issues/604#issuecomment-84051143).

```suggestion
        def __init__(self, value: int):
```

**Comment by stinodego** [comment]
Hm, tough one. I think we want to keep mypy on Python 3.7 in our CI; I'll do some checks.

Maybe we can merge these changes, but keep the flag disabled until we come up with a solution?

File: py-polars/tests/test_df.py (hunk 1 above)
Comment (zundertj) [reply py-polars/tests/test_df.py]: I have changed to `int`, as that is the value used later in the test. I could not drop the `None` though, mypy complains. So leaving that one in.

**Comment by stinodego** [comment]
The Python test CI checks are still not running, it seems.

**Comment by zundertj** [comment]
Tried to fix again, but it seems I am hitting this: https://github.community/t/matrix-cannot-be-used-in-jobs-level-if/17177/12. I.e., this is not possible, as the if-statement is evaluated before the matrix. The only solution I see now is to not use the matrix anymore, and thus duplicate the job effectively for Python 3.7.

Open for anyone (with more GH Actions experience) to suggest a better solution.

**Comment by ritchie46** [comment]
Thanks @zundertj  and @stinodego. Merging it in.

**Comment by ritchie46** [comment]
I had to revert this one because the github action yaml was invalid. We need to run the action if the action file has changed, I shall open an issue for this.

@zundertj can you reopen the PR?

**Pull request merged**